### PR TITLE
CNF-12680:CNF-13488: Add performance profile status for hypershift support

### PR DESCRIPTION
**Pull request by rbaturov**
This PR follows the implementation of this enhancement:
https://github.com/openshift/enhancements/pull/1619

Up until now, there was no defined resolution over how PerformanceProfile
status conditions with the status of the different components created
from the PerformanceProfile (MachineConfig, KubeletConfig, Tuned) are handled and exposed.
To resolve this, a custom ConfigMap is added and will be used as a middleware object.
It will hold the updated status populated by the performance profile controller
and will be watched by the NodePool controller, which will in turn reflect them under NodePool.status.condition.

- [x]  Implementation
- [ ]  E2E config suite adoption

There are E2E tests that are required to be added. we can do so once we adjust the `performanceprofile_controller_test.go` (or add another controller test for the hypershift implementation)

Some additional gaps, that we might address if possible:
* In this implementation, KubeletConfig status is not collected (on hypershift)

### CODE REVIEW

**Comment by Tal-or** [comment]
/retitle CNF-12680: Add performance profile status for hypershift support

**Comment by Tal-or** [comment]
/hold
e2e tests are in progress

**Comment by Tal-or** [comment]
/hold cancel
This PR is reviewable and includes implementation unit-tests and e2e tests

**Comment by Tal-or** [comment]
/retitle: CNF-12680:CNF-13488: Add performance profile status for hypershift support

File: pkg/performanceprofile/controller/performanceprofile/status/status.go [hunk 1]
```diff
@@ -261,3 +266,52 @@ func removeUnMatchedTunedProfiles(nodes []corev1.Node, profiles []tunedv1.Profil
 	}
 	return filteredProfiles
 }
+
+func CalculateUpdatedStatus(prevStatus *performancev2.PerformanceProfileStatus, profileName string, conditions []conditionsv1.Condition) *performancev2.PerformanceProfileStatus {
```
Comment (ffromani) [inline pkg/performanceprofile/controller/performanceprofile/status/status.go]: nit: the name of the function is a bit clunky: `status.CalculateUpdatedStatus` but I don't have good suggestions, so this is totally not-blocking (and can be addressed last, or even later)

Comment (ffromani) [reply pkg/performanceprofile/controller/performanceprofile/status/status.go]: minor: returning an object (not a pointer) signals we are making a copy, especially if we deference it anyway in the call sites; alternatively we can make the receiver an object, not a pointer.

File: pkg/operator/hypershift.go [hunk 2]
```diff
@@ -307,3 +307,7 @@ func generatedConfigMapAnnotations(nodePoolName string) map[string]string {
 		hypershiftNodePoolLabel: nodePoolName,
 	}
 }
+
+func GetTunedUniqueName(tunedName, nodePoolName string) string {
```
Comment (ffromani) [inline pkg/operator/hypershift.go:311]: minor/nit: `Get` Unique Name or `Make` Unique Name?

File: pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	hypershiftconsts "github.com/openshift/cluster-node-tuning-operator/pkg/performanceprofile/controller/performanceprofile/hypershift/consts"
+	"github.com/openshift/cluster-node-tuning-operator/pkg/performanceprofile/controller/performanceprofile/status"
+	conditionsv1 "github.com/openshift/custom-resource-status/conditions/v1"
+)
+
+const (
+	ppStatusConfigMapPrefix = "status"
+)
+
+func getStatusConfigMapName(instanceName string) string {
```
Comment (ffromani) [inline pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go]: minor/nit: `get` or `make` status config map name?

File: pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	key := client.ObjectKeyFromObject(cm)
+	if prevStatusFound {
+		klog.InfoS("Updating status", "ConfigMap", key.String())
+		if err := cli.Update(ctx, cm); err != nil {
+			return fmt.Errorf("failed to update status ConfigMap %q: %w", key.String(), err)
+		}
+		klog.InfoS("Status updated ", "ConfigMap", key.String())
+	} else {
```
Comment (ffromani) [inline pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go:55]: minor/nit: please remove the `else` block and align to the left

File: pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+		klog.InfoS("Creating status", "ConfigMap", key.String())
+		if err := cli.Create(ctx, cm); err != nil {
+			return fmt.Errorf("failed to create ConfigMap %q: %w", key.String(), err)
+		}
+		klog.InfoS("Status created", "ConfigMap", key.String())
+	}
+	return nil
+}
+
+func ppStatusConfigMap(instance *corev1.ConfigMap, profileName string, scheme *runtime.Scheme) (*corev1.ConfigMap, error) {
```
Comment (ffromani) [inline pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go:65]: minor/nit: `makePerformanceProfileStatusConfigMap`?

File: pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	messageString := status.GetTunedProfilesMessage(tunedProfileList.Items)
+	if len(messageString) == 0 {
+		return nil, nil
+	}
+	return status.GetDegradedConditions(status.ConditionReasonTunedDegraded, messageString), nil
+}
+
+func getPreviousStatusFrom(ctx context.Context, cli client.Client, cm *corev1.ConfigMap) (*performancev2.PerformanceProfileStatus, bool, error) {
```
Comment (ffromani) [inline pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go:94]: non-blocking question: is this get cached somehow somewhere (client-go perhaps) right?

File: pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go [hunk 7]
```diff
@@ -235,3 +266,7 @@ func DecodeManifest(b []byte, scheme *runtime.Scheme, obj runtime.Object) (bool,
 	}
 	return false, err
 }
+
+func GetStatusConfigMapName(instanceName string) string {
```
Comment (ffromani) [inline pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go:336]: minor./nit: `get` or `make`?
didn't we have something similar or the same code elsewhere?

File: pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go [hunk 8]
```diff
@@ -32,6 +32,30 @@ type ControlPlaneClientImpl struct {
 	// hostedControlPlaneNamespaceName is the namespace name on the management cluster
 	// on which the control-plane objects reside
 	hostedControlPlaneNamespaceName string
+
```
Comment (ffromani) [inline pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go:35]: this is well intentioned, but two mode of operations create room for bugs. We need to evaluate the opportunity cost here

File: pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go [hunk 9]
```diff
@@ -90,6 +90,14 @@ func (ci *ControlPlaneClientImpl) Update(ctx context.Context, obj client.Object,
 	return ci.Client.Update(ctx, obj, opts...)
 }
 
+func (ci *ControlPlaneClientImpl) Delete(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error {
```
Comment (ffromani) [inline pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go:97]: please add in the commit message a line or few about the intended usage for the `Delete` API

File: pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go [hunk 10]
```diff
@@ -33,7 +33,8 @@ type ControlPlaneClientImpl struct {
 	// on which the control-plane objects reside
 	hostedControlPlaneNamespaceName string
 
-	opts controlPlaneClientImplOptions
+	yamlSerializer *serializer.Serializer
```
Comment (ffromani) [inline pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go:36]: ok, but why? performances?

File: test/e2e/performanceprofile/functests/utils/tuned/tuned.go [hunk 11]
```diff
@@ -3,6 +3,7 @@ package tuned
 import (
 	"context"
 	"fmt"
+	"hash/fnv"
```
Comment (ffromani) [inline test/e2e/performanceprofile/functests/utils/tuned/tuned.go:6]: I'm not against this approach, but there should be another PR which adds the bindata compilation anyway (#1019 ?). Let's sync about this.

**Review by ffromani — COMMENTED** [review]
the hottest topic is the testing mode, which is well intentioned but I'm afraid it can backfire later on. Let's spend more time (let me spend more time) assessing the approach. Roughly half of the changes are straightforward fixes or changes which seem fine already. Will need at least another review pass.

File: pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go (hunk 10 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go:36]: Spare the need of recreating the serializer on every call

File: test/e2e/performanceprofile/functests/utils/tuned/tuned.go (hunk 11 above)
Comment (Tal-or) [reply test/e2e/performanceprofile/functests/utils/tuned/tuned.go:6]: Good. I went with the opposite approach but probably in your case it was a must.
I suggest to apply https://github.com/openshift/cluster-node-tuning-operator/pull/1019/commits/426451af8e24552b16c1c339ddbf04b6469f766d on separate PR and the sooner the better.

File: pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go (hunk 6 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go:94]: Caching when we're having two clients (i.e. two etcds) is something that deserves more intention and deeper investigation. ATM I do not have all the answers unfortunately, but what I'm knowing for sure is that we're caching only the hosted-cluster objects:
https://github.com/openshift/cluster-node-tuning-operator/blob/master/cmd/cluster-node-tuning-operator/main.go#L137

File: pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go (hunk 9 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go:97]: I'm pretty sure I did - from the commit message:
```
The `Delete` call deletes the `ConfigMap` that holds an encoded data of the object that was given for deletion.
```

**Comment by Tal-or** [comment]
/retitle CNF-12680:CNF-13488: Add performance profile status for hypershift support

File: pkg/performanceprofile/controller/performanceprofile/status/status.go (hunk 1 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/status/status.go]: > minor: returning an object (not a pointer) signals we are making a copy, especially if we deference it anyway in the call sites; alternatively we can make the receiver an object, not a pointer.

We can change it to an actual object. The only reason I see for keeping it a pointer is because it's easier to check whether the status got updated or not, by checking if the function returned a `nil`.
We can add additional argument for signaling that tough

File: pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go (hunk 3 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go]: similar to: https://github.com/openshift/cluster-node-tuning-operator/blob/570bbfd7e761611f58ad20f3e478aec1f044a814/pkg/performanceprofile/controller/performanceprofile/components/utils.go#L16

File: pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go (hunk 7 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go:336]: We do, I made this one: https://github.com/openshift/cluster-node-tuning-operator/pull/1089#discussion_r1685683711 public

File: pkg/operator/hypershift.go (hunk 2 above)
Comment (Tal-or) [reply pkg/operator/hypershift.go]: Done. Thanks

File: pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go (hunk 4 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go]: Done. Thank you

File: pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go (hunk 5 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go]: Done

**Comment by Tal-or** [comment]
/hold
Depends on #1116

**Comment by Tal-or** [comment]
> This PR follows the implementation of this enhancement:
https://github.com/openshift/enhancements/pull/1619
>
> Up until now, there was no defined resolution over how PerformanceProfile
status conditions with the status of the different components created
from the PerformanceProfile (MachineConfig, KubeletConfig, Tuned) are handled and exposed.
To resolve this, a custom ConfigMap is added and will be used as a middleware object.
It will hold the updated status populated by the performance profile controller
and will be watched by the NodePool controller, which will in turn reflect them under NodePool.status.condition.
>
> Implementation
 E2E config suite adoption
There are E2E tests that are required to be added. we can do so once we adjust the performanceprofile_controller_test.go (or add another controller test for the hypershift implementation)
>
>Some additional gaps, that we might address if possible:
>
>In this implementation, KubeletConfig status is not collected (on hypershift)

- [x] Implementation
- [x]  E2E config suite adoption 
- [x] performanceprofile_controller_test.go (or add another controller test for the hypershift implementation) 

In this implementation, KubeletConfig status is not collected (on hypershift) - And it probably won't since the status is not reflected by KubeletConfig on Hypershift at all.

**Comment by Tal-or** [comment]
/hold cancel

#1116 merged

File: pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go (hunk 6 above)
Comment (ffromani) [reply pkg/performanceprofile/controller/performanceprofile/hypershift/status/status.go:94]: ok, good enough for now

File: pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go (hunk 8 above)
Comment (ffromani) [reply pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go:35]: after some (admittedly not too much) more reflection, I think this is well intentioned but too risky. Test code is meant to exercise the production code as much as possible and this is a nice idea but make us diverge too much. I'm already a bit concerned about the previous comment (which gets `Status` from configmap and transparently merges into the main object), this one feels far too above the threshold.

I think we should accept the performance cost for the time being.

File: pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go (hunk 9 above)
Comment (ffromani) [reply pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go:97]: you indeed described what the `Delete` call would do, but is not obvious which flow needs `Delete` (and how we managed so far?)

Comment (ffromani) [reply pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go:97]: anyway, not really blocking

File: pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go (hunk 10 above)
Comment (ffromani) [reply pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go:36]: the tradeoff here is between reusing objects, which can possibly have unexpected state, vs recreating every time. The latter seems cheap enough, but the serializer should be stateless, so I'm ok with this change, even though I'm skeptical about the gains here.

File: pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go (hunk 8 above)
Comment (ffromani) [reply pkg/performanceprofile/controller/performanceprofile/hypershift/hypershift.go:35]: wait, I misunderstood the code. This really ties to the previous commit and basically enables the API backward  compatibility to restore the status in the same perfprof object. Let me rethink.

**Review by ffromani — COMMENTED** [review]
ok, please:
- remove the testing mode commit
- improve the doc for the commit adding `Delete` (which code needs it? how we managed till now?)

and LGTM

**Comment by Tal-or** [comment]
> ok, please:
> 
> * remove the testing mode commit
> * improve the doc for the commit adding `Delete` (which code needs it? how we managed till now?)
> 
> and LGTM

Done.
See: https://github.com/openshift/cluster-node-tuning-operator/pull/1089/commits/072a8df06e2e661bd385bd6960bca4b70b229a5c

**Review by ffromani — COMMENTED** [review]
/approve
/lgtm

thanks for the updates and sorry for the churn
